### DISCUSSION TITLE
Is using Materialize as a general cache for app backends a good idea?

### DISCUSSION THREAD

**Discussion by Linksku** [MaterializeInc/materialize#8870] [Q&A]
Hi, I'm a frontend dev and I've been trying to tackle the issue of efficiently using derived data for an app backend, Materialize looks really promising!

Materialize's website focuses on analytics, streams, microservices, etc. Would it be a good idea for a general cache for derived data in a typical app or website? E.g. some scenarios where Materialize could help a lot:

- counting the number of votes/likes
- getting the time of the last reply
- denormalizing joins for faster reads

I'm currently storing some of this data in the database, some of it in cache, and some of it is re-computed for every request. Would it be a good idea to compute all of this using Materialize? It would be much faster to fetch from materialized views than re-computing from the normalized database. In addition, I don't have to worry about manually updating the cached values. I would still have a simple read-through cache on top.

In addition, would reads be faster if I copied the materialized views to Postgres/MySQL/etc, or would it be roughly the same speed to just query Materialize directly?

**Comment by rjnn** [comment]
Yes. This is a great use for Materialize. To keep your architecture as simple as possible, I would recommend using Postgres for all your writes, connecting Materialize to Postgres using [direct postgres replication](https://materialize.com/docs/guides/cdc-postgres/#direct-postgres-source). I'd split your workload into three categories: 

(1) All writes always go directly to postgres. 
(2) Reads that are transactional with writes go directly to postgres. 
(3) All other reads go to Materialize. 

Structure your indexes as minimally as possible with minimal overlap - the only indexes you create in postgres are those that make (1) and (2) faster. The only indexes you create in Materialize are those that make (3) faster.

> In addition, would reads be faster if I copied the materialized views to Postgres/MySQL/etc, or would it be roughly the same speed to just query Materialize directly?

I don't understand this question fully, so I'd appreciate a bit more clarification as to what you mean. However, using this architecture, I would not create any materialized views in Postgres - only in Materialize. Once you hook this up end-to-end, any indexes/materialized views/etc that you create in Postgres are immaterial for Materialize (and vice versa) - in fact, all they do is slow Postgres down, so I'd strip Postgres of as many indexes as possible - the more indexes you can migrate from Postgres to Materialize, the better.

> Materialize's website focuses on analytics, streams, microservices, etc.

This is good criticism. You are right that we don't say more on this kind of use. Perhaps we should.

**Comment by Linksku** [comment]
Thanks for your reply, I will try Materialize this week.

> I don't understand this question fully, so I'd appreciate a bit more clarification as to what you mean.

My understanding is I can use any Postgres client to query the materialized views, but the data isn't stored in Postgres, it's in Materialize's proprietary format. I'd assume Postgres is a bit more optimized for `select`s. So I was thinking a possibility is having Postgres as both a source and a sink. I'd use Materialize to compute the incremental updates, then write the updates to a new Postgres table. I'd query the new Postgres table instead of Materialize and I'd never write to the new table directly.

If it's significantly faster to query Postgres than Materialize, then it could be worth the lag to update the new table. I'll try it myself once I get Materialize set up.

**Reply by rjnn** [reply]
I would not do this (writing back from Materialize -> Postgres) for two reasons: First, we don't support natively writing things back to Postgres, and you would have to do that yourself. That is an error prone process. Second, this just adds end-to-end latency. Materialize _should_ be fast enough for you to query directly. The query driver is also identical - you just have two pg client objects (one for Postgres, one for Materialize) in your server code and you use the appropriate one for the appropriate query. Does that make sense?

**Comment by ruchirK** [comment]
You might also want to look at the [TAIL](https://materialize.com/docs/sql/tail/) command as a way to get the continuous stream of changes to a relation / materialized view, as an alternative to repeated `SELECT`s. We have a [blog post](https://materialize.com/a-simple-and-efficient-real-time-application-powered-by-materializes-tail-command/) that talks through how to build a really simple web application with Materialize + TAIL.
